app/services/course/assessment/submission/auto_grading_service.rb

```diff
@@ -65,4 +66,25 @@ def aggregate_failures(jobs)
     error_messages = failed_jobs.map { |job| job.error['message'] }
     raise SubJobError, error_messages.to_sentence
   end
+
+  def publish_grade(submission)
+    submission.answers.select(&:submitted?).each(&:publish!)
+    submission.points_awarded = calculate_exp(submission).to_i
+    submission.publish!
+  end
+
+  # Calculating scheme:
+  #   Submit in time: ( base_exp + bonus_exp ) * actual_grade / max_grade
+  #   Late submission: base_exp * actual_grade / max_grade
```

Comment:
Late implies that it's after the whole assessment's due date, not just the bonus. How about
"Submit before bonus cutoff" and "Submit after bonus cutoff" ?